### ISSUE TITLE
Unify model format customize string or Json

### ISSUE THREAD

**Issue by tongwu-msft** [microsoft/LightGBM#4887]
## Summary

Currently we have 2 model formats: customize string and json. It is hard to maintain 2 serialize/deserialize for model. Specially for some components with complex parameters like category encoders... (almost 1/3 code is about serialize/deserialize and hard to extend). We'd better only support 1 model format.

## Motivation

Json format is more reasonable to me, it is standard format and easy to handle hierarchy structure. Backward should be a big problem, I would suggest for new features we can use DumpToString() == DumpToJson(), and replace old component to use json as default format and keep customize string as deprecate method.

**Comment by tongwu-msft** [comment]
@jameslamb , @shiyu1994 , @StrikerRUS , @guolinke, not sure your opinion on this? thanks!

**Comment by StrikerRUS** [comment]
I'm for JSON too! However, I guess we should compare performance of these two formats to make a thoughtful choice.

Note, XGBoost is also using JSON as their main serialization format:
https://github.com/dmlc/xgboost/issues/3980
https://github.com/dmlc/xgboost/pull/5046
https://github.com/dmlc/xgboost/pull/5772

**Comment by StrikerRUS** [comment]
Linking original issue: #2604.

**Comment by guolinke** [comment]
I remember these is a multi-threading speed-up for string format model file loading. 
If the loading speed of json format is comparable,  I am okay with the change. 
And we may need to consider backward compatibility.

**Comment by jameslamb** [comment]
@guolinke 's comments make sense to me, that it would be harder to take advantage of parallelism when parsing JSON than when parsing the existing text format. But I think it's worth trying and benchmarking!

I'll also note here that @lemire, author of [`fast_double_parser`](https://github.com/lemire/fast_double_parser) who was so helpful with #3405, has worked on a fast JSON parser that we could consider vendoring in (similarly to `fmt`, `fast_double_parser`, `etc.): https://github.com/simdjson/simdjson

**Comment by lemire** [comment]
@jameslamb We would be there to help if you choose to adopt simdjson. At this point, it is a mature library, with extensive tests and documentation. We support two APIs: our On Demand front-end (high perf. default) and a conventional DOM approach.

**Comment by StrikerRUS** [comment]
Our friends at XGBoost are trying to adopt Universal Binary JSON format as a binary serialization format: https://github.com/dmlc/xgboost/pull/7545.
